### ISSUE TITLE
Parallelizing data transformations

### ISSUE THREAD

**Issue by lqrz** [jpmml/sklearn2pmml#266]
Hi, as part of my pipeline, I'm encoding my categorical variables by making use of the `CategoricalDomain` decorator, and the `PMMLLabelEncoder` preprocessor.

This causes the `DiscreteDomain.fit()._valid_value_mask()` to call the `util.eval_rows()` function, which is applied iterating sequentially row-by-row over the given pd.DataFrame (in this case), which results in a very expensive operation.

In my particular example, I have DataFrames of tens of millions rows, and it's impossible to use the lib.

It would be great to have the `util.eval_rows()` function parallelised.

---

This is my pipeline definition:

```python

model_params = {...}
model = LGBMClassifier(**model_params)

transforms = [
    (
        ['COL_CATEGORICAL_1'],
        [
            CategoricalDomain(
                with_data=True,
                with_statistics=False,
                invalid_value_treatment="as_missing",
                missing_value_treatment="as_value",
                missing_value_replacement=pd.NA,
                dtype=str),
            PMMLLabelEncoder(missing_values=99999)
        ]
    ),
    (
        ['COL_CONTINUOUS_1'],
        [
            ContinuousDomain(
                with_data=True,
                with_statistics=False,
                invalid_value_treatment="as_is",
                missing_value_treatment="return_invalid",
                dtype=float)
        ]
    )
]

preprocessing_pipeline = DataFrameMapper(
    features=transforms,
    sparse=False,
    df_out=True,
    input_df=True,
    drop_cols=None
)

# instantiate training pipeline
pipeline = PMMLPipeline([
    ('preprocess', preprocessing_pipeline),
    ('classifier', model)
])
```

---

I forked the lib and added a few logging steps. As it can be seen (im running a small dataset in this example), this is the only bottleneck op in my pipeline:

![image](https://user-images.githubusercontent.com/8641479/109034695-30407700-76c8-11eb-8f66-7cf4196a9588.png)

![image](https://user-images.githubusercontent.com/8641479/109036144-9a0d5080-76c9-11eb-9b3c-cfb784129e40.png)

![image](https://user-images.githubusercontent.com/8641479/109036163-9ed20480-76c9-11eb-8c5b-cbf6bf6372c6.png)

---
Environment info:
* sklearn2pmml version: 0.56.2
* Python: 3.7.9
* OS: Debian GNU/Linux 10 (buster)

**Comment by vruusmann** [comment]
> It would be great to have the util.eval_rows() function parallelised.

In principle, that's a great idea. Can you make the `util.eval_rows()` function run in parallel? I probably can't, because Numpy/Scikit-Learn/Python is not my primary language.

Things you could try:

- Just leave out the initial decorator step. It's not needed for anything critical (the whole Scikit-Learn framework is doing very nicely without decorators). The subsequent `(PMML)LabelBinarizer` step will pick up valid category levels, and the resulting PMML document should look and behave effectively the same.

- Toggle some `CategoricalDomain` init parameters. Valid value checks are only performed when needed.

- Replace `sklearn_pandas.DataFrameMapper` with `sklearn.compose.ColumnTransformer`. The column transformer has the `n_jobs` attribute, and is therefore able to run different column pre-processors in parallel.

> sklearn2pmml version: 0.56.2

Seriously?

Perhaps this issue has been resolved in newer SkLearn2PMML package versions...

**Comment by vruusmann** [comment]
> Replace sklearn_pandas.DataFrameMapper with sklearn.compose.ColumnTransformer. The column transformer has the n_jobs attribute, and is therefore able to run different column pre-processors in parallel.

This is the suggested workaround! 

You should be preparing multiple columns in parallel, using `sklearn.compose.ColumnTransformer(n_jobs = 4)`.

Controlling the parallelization at the parent/top-level meta-transformer level (eg. ColumnTransformer`) is technically superior to controlling the same at child/nested transformer levels (eg. `CategoricalDomain`).

**Comment by lqrz** [comment]
Thanks @vruusmann for your quick response!

I ended up modifying the `DiscreteDomain._valid_value_mask()` method so that instead of iterating row-by-row on a pandas Series or DataFrame, it uses its internal -vectorised- `.isin()` implementation, yielding a great performance boost:

```python
def _valid_value_mask(self, X, where):

    if hasattr(self, "data_"):

        def is_valid(x): return x.isin(self.data_) if hasattr(x, "isin") else x in self.data_

        mask = X.isin(self.data_).values if isinstance(X, pandas.Series) or isinstance(X, pandas.DataFrame) else eval_rows(X, is_valid, dtype=bool)

        mask = (numpy.asarray(mask, dtype=bool)).reshape(X.shape)

        return numpy.logical_and(mask, where)

    return super(DiscreteDomain, self)._valid_value_mask(X, where)
```

On a dataframe of ~180K rows, my preprocessing transformation goes from:
- [Legacy code]: 1min 50s ± 181 ms
- [Vectorised]: 614 ms ± 2.11 ms

![image](https://user-images.githubusercontent.com/8641479/109187460-38fd7f80-7792-11eb-82ab-5ca7d5b536ef.png)

![image](https://user-images.githubusercontent.com/8641479/109187423-31d67180-7792-11eb-9eb2-cec39a4e778d.png)

Also notice that, in the legacy code, inside `util.eval_rows()` there is a call to `return X.apply(func, axis=1)` for both pandas DataFrames and Series, when the pandas.DataFrame signature is different from the pandas.Series one (i.e. the latter does not accept the `axis=` kwarg). Thus, if I define my transformations as single column (DataFrameMapper passes a pandas.Series), and the process crashes.
In the timeit images above, I'm not getting the array shape warning cause I'm defining the transform as single-col (and now the `apply` function is no longer called).

Of course this only applies cause the "invalid mask" definition is asserting the inclusion of an element in a list, for which there is a pandas vectorised function. If in the future this condition changes, I will have to fall back to an `.apply()` approach.

**Comment by vruusmann** [comment]
@lqrz Thanks for prototyping and submitting some actionable advice!

Reopening this issue, now that there's a clear idea what to change about the existing Python code.
